// lib0/dist/encoding.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./binary-1b5a1494.cjs');
require('./map-28a001c9.cjs');
require('./string-8d09325b.cjs');
require('./conditions-fb475c70.cjs');
require('./storage.cjs');
require('./environment-4f8c4c3a.cjs');
require('./buffer-f53235d9.cjs');
require('./math-7cb24cbe.cjs');
require('./number-f4d94318.cjs');
var encoding = require('./encoding-58f13e82.cjs');



exports.Encoder = encoding.Encoder;
exports.createEncoder = encoding.createEncoder;
exports.length = encoding.length;
exports.set = encoding.set;
exports.setUint16 = encoding.setUint16;
exports.setUint32 = encoding.setUint32;
exports.setUint8 = encoding.setUint8;
exports.toUint8Array = encoding.toUint8Array;
exports.write = encoding.write;
exports.writeAny = encoding.writeAny;
exports.writeBigInt64 = encoding.writeBigInt64;
exports.writeBigUint64 = encoding.writeBigUint64;
exports.writeBinaryEncoder = encoding.writeBinaryEncoder;
exports.writeFloat32 = encoding.writeFloat32;
exports.writeFloat64 = encoding.writeFloat64;
exports.writeOnDataView = encoding.writeOnDataView;
exports.writeUint16 = encoding.writeUint16;
exports.writeUint32 = encoding.writeUint32;
exports.writeUint8 = encoding.writeUint8;
exports.writeUint8Array = encoding.writeUint8Array;
exports.writeVarInt = encoding.writeVarInt;
exports.writeVarString = encoding.writeVarString;
exports.writeVarUint = encoding.writeVarUint;
exports.writeVarUint8Array = encoding.writeVarUint8Array;
