// lib0/dist/logging.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./array-5c116aa0.cjs');
require('./map-28a001c9.cjs');
require('./string-8d09325b.cjs');
require('./conditions-fb475c70.cjs');
require('./storage.cjs');
require('./environment-4f8c4c3a.cjs');
require('./pair-ab022bc3.cjs');
require('./dom-ed0e5f03.cjs');
require('./object-dcdd6eed.cjs');
require('./function-040981e4.cjs');
require('./json-092190a1.cjs');
require('./math-7cb24cbe.cjs');
require('./eventloop-d21dd5ae.cjs');
require('./metric.cjs');
require('./time-144a4883.cjs');
require('./symbol-c5caa724.cjs');
var logging = require('./logging-9cad5fc9.cjs');



exports.BLUE = logging.BLUE;
exports.BOLD = logging.BOLD;
exports.GREEN = logging.GREEN;
exports.GREY = logging.GREY;
exports.ORANGE = logging.ORANGE;
exports.PURPLE = logging.PURPLE;
exports.RED = logging.RED;
exports.UNBOLD = logging.UNBOLD;
exports.UNCOLOR = logging.UNCOLOR;
exports.VConsole = logging.VConsole;
exports.createModuleLogger = logging.createModuleLogger;
exports.createVConsole = logging.createVConsole;
exports.group = logging.group;
exports.groupCollapsed = logging.groupCollapsed;
exports.groupEnd = logging.groupEnd;
exports.print = logging.print;
exports.printCanvas = logging.printCanvas;
exports.printDom = logging.printDom;
exports.printError = logging.printError;
exports.printImg = logging.printImg;
exports.printImgBase64 = logging.printImgBase64;
exports.vconsoles = logging.vconsoles;
exports.warn = logging.warn;
