// @sammacbeth/dat-shared-doc/build/cjs/metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const yjs_1 = require("yjs");
const events_1 = require("events");
class Metadata extends events_1.EventEmitter {
    constructor(doc) {
        super();
        this.doc = doc;
        this.feeds = doc.getMap('feeds');
        this.feeds.observeDeep((event) => {
            this.emit('changed');
        });
    }
    addFeed(kind, key, properties = {}) {
        this.doc.transact(() => {
            if (!this.feeds.has(kind)) {
                this.feeds.set(kind, new yjs_1.Map());
            }
            const feedsOfKind = this.feeds.get(kind);
            if (feedsOfKind.has(key)) {
                throw new Error('feed already exists');
            }
            feedsOfKind.set(key, properties);
        });
    }
    getFeeds(kind) {
        if (!this.feeds.has(kind)) {
            return null;
        }
        return this.feeds.get(kind);
    }
    getFeedProperties(kind, key) {
        const feeds = this.getFeeds(kind);
        if (!feeds || !feeds.has(key)) {
            return null;
        }
        return feeds.get(key);
    }
}
exports.default = Metadata;
